File: angular/src/app/models/product.ts
import {Newproduct} from './requests/newproduct';

export class Discount {
  absolute: boolean;
  id : string;
  amount : number;
  reason : string;
  constructor(){}
}

export class Product{
  id :  string;
  name :  string;
  price : number;
  priceWithDiscount : number;
  amount;
  discount : {
    absolute: boolean;
    id : string;
    amount : number;
    reason : string
  };
  constructor(){
    this.discount = new Discount();
  }

  tojson(){
    return {

    }
  }
  generate(np: Newproduct): Product{
    this.name = np.name;
    this.price = np.price;
    if(np.discountAmount != 0){
      this.discount.amount = np.discountAmount;
      this.discount.reason = np.reason;
      this.discount.absolute = np.discountIsAbsolute;
    }

    return this;
  }
}

File: Backend/Backend/Domain/Models/Product.cs
using Backend.Domain.Requests;
using MongoDB.Bson;
using MongoDB.Bson.Serialization.Attributes;

namespace Backend.Domain.Models
{
    public class Product
    {
        [BsonId] [BsonRepresentation(BsonType.ObjectId)]
        public string Id; 
        
        public string Name;

        public double Price;

        public double PriceWithDiscount;

        public IDiscount Discount;

        public void calculatePrice()
        {
            PriceWithDiscount = Discount.CalculateNewPrice(Price);
        }

        public Product()
        {
        }

        public Product(NewProductRequest req)
        {
            if(req.Id != null)
                Id = req.Id;
            Name = req.Name;
            Price = req.Price;
            if (!req.DiscountAmount.Equals(0))
            {
                IDiscount disc;
                if(req.DiscountIsAbsolute)
                    disc = new AbsoluteDiscount();
                else
                    disc = new PercentageDiscount();
                disc.Amount = req.DiscountAmount;
                disc.Reason = req.Reason;
                Discount = disc;
                calculatePrice();
            }
        }
    }
}